iwyu.cc

```diff
@@ -3779,13 +3781,20 @@ class IwyuAstConsumer
     if (const ReferenceType* reftype = DynCastFrom(arg_type)) {
       arg_type = reftype->getPointeeTypeAsWritten().getTypePtr();
     }
-    if (!IsTemplatizedType(arg_type))
+    const TemplateSpecializationType* arg_tmpl = DynCastFrom(arg_type);
+    if (arg_tmpl == nullptr)
       return Base::VisitUnaryExprOrTypeTraitExpr(expr);
 
     const map<const Type*, const Type*> resugar_map
         = GetTplTypeResugarMapForClass(arg_type);
-    instantiated_template_visitor_.ScanInstantiatedType(
-        arg_type, current_ast_node(), resugar_map);
+
+    //Special case: We are instantiating the type in the context of an
+    //expression. Need to push the type to the AST stack explicitly.
+    ASTNode* out_node = current_ast_node();
```

Comment:
I am not sure I understand what you mean. We need to push one node to the AST stack before entering the *ScanInstantiatedType* because it expects that the top is a *TemplateSpecializationType*. 

Something I realized during inspection is that I don't need to use the *CurrentASTNodeUpdater* to restore the parent after the template visitor is done. Now it is down to two lines :)

Reply:
`CurrentASTNodeUpdater` "out here" protects the visitor's `current_ast_node`. But the `instantiated_template_visitor_` holds its own `ASTNode`, and I'm guessing you'd need to save that (using e.g. `CurrentASTNodeUpdater`) so you get back to a known state after `ScanInstantiatedType`. Since the `instantiated_template_visitor_` is long-lived, its state might be worth preserving.

I'm saying this as someone who knows very little about the details here, though, so take with a grain of salt.

Reply:
Exactly, CurrentASTNodeUpdate modifies current_ast_node upon construction and restores it upon destruction, but in this case I don't need to modify it or to restore it. I just want to push a new ASTNode on the stack and pass it to the template visitor. That's why in the changed code I don't use it anymore.